### ISSUE TITLE
Question: Programming language choice

### ISSUE THREAD

**Issue by ovalseven8** [saltyrtc/saltyrtc-client-rs#32]
Probably the wrong place to ask, but I've not found a better one. :/

Actually, I have just a short question out of curiosity. As far as I understand, this library is mainly for the iOS app of Threema.

Just wanted to ask why Rust has been chosen for this library and not Golang or Objective-C? I do not ask because I think one is better than the other (I have no Rust, Go or Objective-C experience), just because of curiosity. While following the development, I have however seen that Rust has some problems because of lacking stable crates. Golang is very good at networking and has stable libraries.

**Comment by dbrgn** [comment]
> As far as I understand, this library is mainly for the iOS app of Threema.

This library is mainly a native implementation of a SaltyRTC client. Yes, it might also be used for Threema on iOS.

Go is not very easy to integrate with code written in other programming languages. Furthermore it has a GC which introduces a host of memory ownership problems when binding to a language that also has a GC.

Objective-C mainly (only?) runs on Apple devices and is probably obsolete by now. Swift is much more modern (and quite influenced by Rust btw), but it can also not really be used outside of the Apple ecosystem.

When writing a library that should be easily integrateable with any codebase that can talk to C, then traditionally that was done in C or C++. Nowadays, Rust brings us a lot of advantages, while still being very easy to integrate with other C code.

In other words, by writing a SaltyRTC client in Rust, we can use it from any language that can talk to C.

If you're interested in the advantages of Rust itself, I can recommend the free ebook [Why Rust](http://www.oreilly.com/programming/free/why-rust.csp) by O'Reilly.

**Comment by dbrgn** [comment]
PS: Feel free to join our [Gitter channel](https://gitter.im/saltyrtc/Lobby) for general questions. There's also an IRC bridge.

**Comment by ovalseven8** [comment]
Thanks a lot, learned something again. :)
